In [4]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [6]:
file_path_historical = "/content/drive/MyDrive/Pt/Data/historical_data.csv"
file_path_fear_greed = "/content/drive/MyDrive/Pt/Data/fear_greed_index.csv"
df_history = pd.read_csv(file_path_historical)
df_fear_greed = pd.read_csv(file_path_fear_greed)

In [7]:
df_history.head()

Account  Coin  Execution Price  \
0  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9769   
1  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9800   
2  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9855   
3  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9874   
4  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9894   

   Size Tokens  Size USD Side     Timestamp IST  Start Position Direction  \
0       986.87   7872.16  BUY  02-12-2024 22:50        0.000000       Buy   
1        16.00    127.68  BUY  02-12-2024 22:50      986.524596       Buy   
2       144.09   1150.63  BUY  02-12-2024 22:50     1002.518996       Buy   
3       142.98   1142.04  BUY  02-12-2024 22:50     1146.558564       Buy   
4         8.73     69.75  BUY  02-12-2024 22:50     1289.488521       Buy   

   Closed PnL                                   Transaction Hash     Order ID  \
0         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
1         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
2         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
3         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
4         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   

   Crossed       Fee      Trade ID     Timestamp  
0     True  0.345404  8.950000e+14  1.730000e+12  
1     True  0.005600  4.430000e+14  1.730000e+12  
2     True  0.050431  6.600000e+14  1.730000e+12  
3     True  0.050043  1.080000e+15  1.730000e+12  
4     True  0.003055  1.050000e+15  1.730000e+12

In [8]:
df_fear_greed.head()

timestamp  value classification        date
0  1517463000     30           Fear  2018-02-01
1  1517549400     15   Extreme Fear  2018-02-02
2  1517635800     40           Fear  2018-02-03
3  1517722200     24   Extreme Fear  2018-02-04
4  1517808600     11   Extreme Fear  2018-02-05

In [9]:
df_history.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 211224 entries, 0 to 211223
Data columns (total 16 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Account           211224 non-null  object 
 1   Coin              211224 non-null  object 
 2   Execution Price   211224 non-null  float64
 3   Size Tokens       211224 non-null  float64
 4   Size USD          211224 non-null  float64
 5   Side              211224 non-null  object 
 6   Timestamp IST     211224 non-null  object 
 7   Start Position    211224 non-null  float64
 8   Direction         211224 non-null  object 
 9   Closed PnL        211224 non-null  float64
 10  Transaction Hash  211224 non-null  object 
 11  Order ID          211224 non-null  int64  
 12  Crossed           211224 non-null  bool   
 13  Fee               211224 non-null  float64
 14  Trade ID          211224 non-null  float64
 15  Timestamp         211224 non-null  float64
dtypes: bool(1), float64(

In [10]:
df_fear_greed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2644 entries, 0 to 2643
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   timestamp       2644 non-null   int64 
 1   value           2644 non-null   int64 
 2   classification  2644 non-null   object
 3   date            2644 non-null   object
dtypes: int64(2), object(2)
memory usage: 82.8+ KB


In [11]:
df_history.isnull().sum()

Account             0
Coin                0
Execution Price     0
Size Tokens         0
Size USD            0
Side                0
Timestamp IST       0
Start Position      0
Direction           0
Closed PnL          0
Transaction Hash    0
Order ID            0
Crossed             0
Fee                 0
Trade ID            0
Timestamp           0
dtype: int64

In [12]:
df_fear_greed.isnull().sum()

timestamp         0
value             0
classification    0
date              0
dtype: int64

In [13]:
df_his = df_history.copy()
df_fear = df_fear_greed.copy()

# Timestamp Correction

In [14]:
df_his['new_Timestamp'] = pd.to_datetime(df_his['Timestamp IST'], dayfirst = True)
df_his['new_date'] = pd.to_datetime(df_his['new_Timestamp'].dt.date)
df_his

Account      Coin  Execution Price  \
0       0xae5eacaf9c6b9111fd53034a602c192a04e082ed      @107           7.9769   
1       0xae5eacaf9c6b9111fd53034a602c192a04e082ed      @107           7.9800   
2       0xae5eacaf9c6b9111fd53034a602c192a04e082ed      @107           7.9855   
3       0xae5eacaf9c6b9111fd53034a602c192a04e082ed      @107           7.9874   
4       0xae5eacaf9c6b9111fd53034a602c192a04e082ed      @107           7.9894   
...                                            ...       ...              ...   
211219  0x72743ae2822edd658c0c50608fd7c5c501b2afbd  FARTCOIN           1.1010   
211220  0x72743ae2822edd658c0c50608fd7c5c501b2afbd  FARTCOIN           1.1010   
211221  0x72743ae2822edd658c0c50608fd7c5c501b2afbd  FARTCOIN           1.1010   
211222  0x72743ae2822edd658c0c50608fd7c5c501b2afbd  FARTCOIN           1.1010   
211223  0x72743ae2822edd658c0c50608fd7c5c501b2afbd  FARTCOIN           1.1010   

        Size Tokens  Size USD  Side     Timestamp IST  Start Position  \
0            986.87   7872.16   BUY  02-12-2024 22:50        0.000000   
1             16.00    127.68   BUY  02-12-2024 22:50      986.524596   
2            144.09   1150.63   BUY  02-12-2024 22:50     1002.518996   
3            142.98   1142.04   BUY  02-12-2024 22:50     1146.558564   
4              8.73     69.75   BUY  02-12-2024 22:50     1289.488521   
...             ...       ...   ...               ...             ...   
211219       382.20    420.80  SELL  25-04-2025 15:35     7546.600000   
211220      2124.10   2338.63  SELL  25-04-2025 15:35     7164.400000   
211221       423.40    466.16  SELL  25-04-2025 15:35     5040.300000   
211222      3599.80   3963.38  SELL  25-04-2025 15:35     4616.900000   
211223      1017.10   1119.83  SELL  25-04-2025 15:35     1017.100000   

         Direction  Closed PnL  \
0              Buy      0.0000   
1              Buy      0.0000   
2              Buy      0.0000   
3              Buy      0.0000   
4              Buy      0.0000   
...            ...         ...   
211219  Close Long    -20.2566   
211220  Close Long   -112.5773   
211221  Close Long    -22.4402   
211222  Close Long   -190.7894   
211223  Close Long    -53.9063   

                                         Transaction Hash     Order ID  \
0       0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
1       0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
2       0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
3       0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
4       0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
...                                                   ...          ...   
211219  0xcd339c08dc7b615a993c0422374d8e02027400092bc2...  88803313862   
211220  0x29e8ede2a3a37aa0eac00422374d8e02029b00ac9f3c...  88803313862   
211221  0x0780085b0c0a943eea800422374d920204c100edf579...  88803313862   
211222  0x349c29934913b25c89e20422374d920204cd008b8a0e...  88803313862   
211223  0xac77fab973c455d77a670422374d9602039800f1f78c...  88803313862   

        Crossed       Fee      Trade ID     Timestamp       new_Timestamp  \
0          True  0.345404  8.950000e+14  1.730000e+12 2024-12-02 22:50:00   
1          True  0.005600  4.430000e+14  1.730000e+12 2024-12-02 22:50:00   
2          True  0.050431  6.600000e+14  1.730000e+12 2024-12-02 22:50:00   
3          True  0.050043  1.080000e+15  1.730000e+12 2024-12-02 22:50:00   
4          True  0.003055  1.050000e+15  1.730000e+12 2024-12-02 22:50:00   
...         ...       ...           ...           ...                 ...   
211219    False  0.042080  1.990000e+14  1.750000e+12 2025-04-25 15:35:00   
211220    False  0.233863  9.260000e+14  1.750000e+12 2025-04-25 15:35:00   
211221    False  0.046616  6.930000e+14  1.750000e+12 2025-04-25 15:35:00   
211222    False  0.396337  4.180000e+14  1.750000e+12 2025-04-25 15:35:00   
211223    False  0.111982  3.270000e+14  1.75000

In [15]:
df_his.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 211224 entries, 0 to 211223
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   Account           211224 non-null  object        
 1   Coin              211224 non-null  object        
 2   Execution Price   211224 non-null  float64       
 3   Size Tokens       211224 non-null  float64       
 4   Size USD          211224 non-null  float64       
 5   Side              211224 non-null  object        
 6   Timestamp IST     211224 non-null  object        
 7   Start Position    211224 non-null  float64       
 8   Direction         211224 non-null  object        
 9   Closed PnL        211224 non-null  float64       
 10  Transaction Hash  211224 non-null  object        
 11  Order ID          211224 non-null  int64         
 12  Crossed           211224 non-null  bool          
 13  Fee               211224 non-null  float64       
 14  Trad

In [16]:
df_fear['new_date'] = pd.to_datetime(df_fear['date'])
df_fear.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2644 entries, 0 to 2643
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   timestamp       2644 non-null   int64         
 1   value           2644 non-null   int64         
 2   classification  2644 non-null   object        
 3   date            2644 non-null   object        
 4   new_date        2644 non-null   datetime64[ns]
dtypes: datetime64[ns](1), int64(2), object(2)
memory usage: 103.4+ KB


# Merged Datasets

In [17]:
df_merged = pd.merge(df_fear,df_his,on='new_date')
df_merged

timestamp  value classification        date   new_date  \
0       1682919000     63          Greed  2023-05-01 2023-05-01   
1       1682919000     63          Greed  2023-05-01 2023-05-01   
2       1682919000     63          Greed  2023-05-01 2023-05-01   
3       1701754200     75  Extreme Greed  2023-12-05 2023-12-05   
4       1701754200     75  Extreme Greed  2023-12-05 2023-12-05   
...            ...    ...            ...         ...        ...   
211213  1746077400     53        Neutral  2025-05-01 2025-05-01   
211214  1746077400     53        Neutral  2025-05-01 2025-05-01   
211215  1746077400     53        Neutral  2025-05-01 2025-05-01   
211216  1746077400     53        Neutral  2025-05-01 2025-05-01   
211217  1746077400     53        Neutral  2025-05-01 2025-05-01   

                                           Account Coin  Execution Price  \
0       0x3998f134d6aaa2b6a5f723806d00fd2bbbbce891  ETH           1897.9   
1       0x3998f134d6aaa2b6a5f723806d00fd2bbbbce891  ETH           1897.9   
2       0x3998f134d6aaa2b6a5f723806d00fd2bbbbce891  ETH           1898.6   
3       0xb1231a4a2dd02f2276fa3c5e2a2f3436e6bfed23  ETH           2230.1   
4       0xb1231a4a2dd02f2276fa3c5e2a2f3436e6bfed23  ETH           2230.1   
...                                            ...  ...              ...   
211213  0xbd5fead7180a9c139fa51a103cb6a2ce86ddb5c3  ETH           1805.6   
211214  0xbd5fead7180a9c139fa51a103cb6a2ce86ddb5c3  ETH           1805.6   
211215  0xbd5fead7180a9c139fa51a103cb6a2ce86ddb5c3  ETH           1805.6   
211216  0xbd5fead7180a9c139fa51a103cb6a2ce86ddb5c3  ETH           1805.6   
211217  0xbd5fead7180a9c139fa51a103cb6a2ce86ddb5c3  ETH           1805.6   

        Size Tokens  Size USD  ... Start Position   Direction  Closed PnL  \
0            0.0967    183.53  ...         0.0000   Open Long         0.0   
1            0.0824    156.39  ...         0.0967   Open Long         0.0   
2            0.0722    137.08  ...         0.1791   Open Long         0.0   
3            0.4933   1100.11  ...         0.0000   Open Long         0.0   
4            0.8681   1935.95  ...         0.4933   Open Long         0.0   
...             ...       ...  ...            ...         ...         ...   
211213       2.5000   4514.00  ...      -473.9880  Open Short         0.0   
211214       5.0000   9028.00  ...      -476.4880  Open Short         0.0   
211215       8.8000  15889.28  ...      -481.4880  Open Short         0.0   
211216      14.1217  25498.14  ...      -490.2880  Open Short         0.0   
211217       5.4903   9913.29  ...      -504.4097  Open Short         0.0   

                                         Transaction Hash     Order ID  \
0       0x875d3e1af52b5b758e4f04015b774e0111006a118601...    173271100   
1       0x875d3e1af52b5b758e4f04015b774e0111006a118601...    173271100   
2       0x875d3e1af52b5b758e4f04015b774e0111006a118601...    173271100   
3       0x02721e3c6061bddfdc1204065f9dfe0138003649ef8c...   4064960965   
4       0x02721e3c6061bddfdc1204065f9dfe0138003649ef8c...   4064960965   
...                                                   ...          ...   
211213  0xb979fffcb79441fdd76404228f73930202a600839132...  90068293884   
211214  0xb979fffcb79441fdd76404228f73930202a600839132...  90068293884   
211215  0xb979fffcb79441fdd76404228f73930202a600839132...  90068293884   
211216  0xb979fffcb79441fdd76404228f73930202a600839132...  90068293884   
211217  0xb979fffcb79441fdd76404228f73930202a600839132...  90068293884   

       Crossed       Fee      Trade ID     Timestamp       new_Timestamp  
0         True  0.000000  0.000000e+00  1.680000e+12 2023-05-01 01:06:00  
1         True  0.000000  0.000000e+00  1.680000e+12 2023-05-01 01:06:00  
2         True  0.000000  0.000000e+00  1.680000e+12 2023-05-01 01:06:00  
3         True  0.275027  6.320000e+14  1.700000e+12 2023-12-05 03:11:00  
4         True  0.483987  2.340000e+14  1.700000e+12 2023-12-05 03:11:00  
...        ...       ...   

# Daily PnL per trader (or per account)

In [18]:
daily_pnl = df_merged.groupby(['Account','date'])['Closed PnL'].sum()

In [19]:
daily_pnl = daily_pnl.reset_index()

In [20]:
daily_pnl.head(20)

Account        date     Closed PnL
0   0x083384f897ee0f19899168e3b1bec365f52a9012  2024-11-11       0.000000
1   0x083384f897ee0f19899168e3b1bec365f52a9012  2024-11-17       0.000000
2   0x083384f897ee0f19899168e3b1bec365f52a9012  2024-11-18       0.000000
3   0x083384f897ee0f19899168e3b1bec365f52a9012  2024-11-22  -21227.000000
4   0x083384f897ee0f19899168e3b1bec365f52a9012  2024-11-26    1603.100000
5   0x083384f897ee0f19899168e3b1bec365f52a9012  2024-11-28 -132271.000000
6   0x083384f897ee0f19899168e3b1bec365f52a9012  2024-11-30       0.000000
7   0x083384f897ee0f19899168e3b1bec365f52a9012  2024-12-06 -175611.000056
8   0x083384f897ee0f19899168e3b1bec365f52a9012  2025-01-19   71393.964098
9   0x083384f897ee0f19899168e3b1bec365f52a9012  2025-01-20   40218.599998
10  0x083384f897ee0f19899168e3b1bec365f52a9012  2025-02-02   76710.000000
11  0x083384f897ee0f19899168e3b1bec365f52a9012  2025-02-03  101011.685664
12  0x083384f897ee0f19899168e3b1bec365f52a9012  2025-02-04  375620.270243
13  0x083384f897ee0f19899168e3b1bec365f52a9012  2025-02-05   44400.905800
14  0x083384f897ee0f19899168e3b1bec365f52a9012  2025-02-09   30897.088700
15  0x083384f897ee0f19899168e3b1bec365f52a9012  2025-02-10   55086.000000
16  0x083384f897ee0f19899168e3b1bec365f52a9012  2025-02-11       0.000000
17  0x083384f897ee0f19899168e3b1bec365f52a9012  2025-02-24   10152.122240
18  0x083384f897ee0f19899168e3b1bec365f52a9012  2025-02-25  115445.395538
19  0x083384f897ee0f19899168e3b1bec365f52a9012  2025-02-28  115287.000000

In [21]:
daily_stats = df_merged.groupby(['Account', 'date']).agg({
    'Closed PnL': 'sum',
    'Order ID': 'nunique',
    'Size USD': 'mean'
}).reset_index()
daily_stats.columns = ['Account', 'date', 'Total_PnL', 'Trade_Count', 'Avg_Trade_Size']

In [22]:
daily_stats.head(10)

Account        date      Total_PnL  \
0  0x083384f897ee0f19899168e3b1bec365f52a9012  2024-11-11       0.000000   
1  0x083384f897ee0f19899168e3b1bec365f52a9012  2024-11-17       0.000000   
2  0x083384f897ee0f19899168e3b1bec365f52a9012  2024-11-18       0.000000   
3  0x083384f897ee0f19899168e3b1bec365f52a9012  2024-11-22  -21227.000000   
4  0x083384f897ee0f19899168e3b1bec365f52a9012  2024-11-26    1603.100000   
5  0x083384f897ee0f19899168e3b1bec365f52a9012  2024-11-28 -132271.000000   
6  0x083384f897ee0f19899168e3b1bec365f52a9012  2024-11-30       0.000000   
7  0x083384f897ee0f19899168e3b1bec365f52a9012  2024-12-06 -175611.000056   
8  0x083384f897ee0f19899168e3b1bec365f52a9012  2025-01-19   71393.964098   
9  0x083384f897ee0f19899168e3b1bec365f52a9012  2025-01-20   40218.599998   

   Trade_Count  Avg_Trade_Size  
0            2     5089.718249  
1            3     7976.664412  
2            1    23734.500000  
3            1    28186.666667  
4            2    17248.148148  
5            2    16058.536341  
6            1    48561.685652  
7            1    36736.362424  
8           12    11410.954022  
9            2     9400.588039

In [23]:
Total_PnL = df_merged.groupby(['classification']).agg({
    'Closed PnL': 'sum',
    'Order ID': 'nunique',
    'Size USD': 'mean'
})

In [24]:
Total_PnL.head(10)

Closed PnL  Order ID     Size USD
classification                                     
Extreme Fear    7.391102e+05      2880  5349.731843
Extreme Greed   2.715171e+06     15708  3112.251565
Fear            3.357155e+06      8979  7816.109931
Greed           2.150129e+06     15099  5736.884375
Neutral         1.292921e+06      7890  4782.732661

# Win Rate

In [25]:
df_merged['is_win'] = df_merged['Closed PnL'] > 0
win_rate_df = df_merged.groupby('Account')['is_win'].mean() * 100
win_rate_df

Account
0x083384f897ee0f19899168e3b1bec365f52a9012    35.961236
0x23e7a7f8d14b550961925fbfdaa92f5d195ba5bd    44.271978
0x271b280974205ca63b716753467d5a371de622ab    30.191651
0x28736f43f1e871e6aa8b1148d38d4994275d72c4    43.858463
0x2c229d22b100a7beb69122eed721cee9b24011dd    51.991355
0x3998f134d6aaa2b6a5f723806d00fd2bbbbce891    45.521472
0x39cef799f8b69da1995852eea189df24eb5cae3c    32.766787
0x3f9a0aadc7f04a7c9d75dc1b5a6ddd6e36486cf6    47.590361
0x420ab45e0bd8863569a5efbb9c05d91f40624641    23.498695
0x430f09841d65beb3f27765503d0f850b8bce7713    48.423605
0x47add9a56df66b524d5e2c1993a43cde53b6ed85    35.344524
0x4acb90e786d897ecffb614dc822eb231b4ffb9f4    48.622590
0x4f93fead39b70a1824f981a54d4e55b278e9f760    36.036392
0x513b8629fe877bb581bf244e326a047b249c4ff1    40.119320
0x6d6a4b953f202f8df5bed40692e7fd865318264a    43.179487
0x72743ae2822edd658c0c50608fd7c5c501b2afbd    34.591195
0x72c6a4624e1dffa724e6d00d64ceae698af892a0    30.337079
0x75f7eeb85dc639d5e99c78f95393aa9a5f1170d4    81.087638
0x7f4f299f74eec87806a830e3caa9afa5f2b9db8f    45.221296
0x8170715b3b381dffb7062c0298972d4727a0a63b    38.274288
0x8381e6d82f1affd39a336e143e081ef7620a3b7f    33.019362
0x8477e447846c758f5a675856001ea72298fd9cb5    26.196826
0x92f17e8d81a944691c10e753af1b1baae1a2cd0d    28.538663
0xa0feb3725a9335f49874d7cd8eaad6be45b27416    34.578661
0xa520ded057a32086c40e7dd6ed4eb8efb82c00e0    57.314149
0xae5eacaf9c6b9111fd53034a602c192a04e082ed    40.852575
0xaf40fdc468c30116bd3307bcbf4a451a7ebf1deb    32.209738
0xb1231a4a2dd02f2276fa3c5e2a2f3436e6bfed23    33.713432
0xb899e522b5715391ae1d4f137653e7906c5e2115    43.861100
0xbaaaf6571ab7d571043ff1e313a9609a10637864    46.758211
0xbd5fead7180a9c139fa51a103cb6a2ce86ddb5c3    32.752745
0xbee1707d6b44d4d52bfe19e41f8a828645437aab    42.823014
Name: is_win, dtype: float64

In [26]:
direction_map = {
    'Open Long': 'Long', 'Close Long': 'Long', 'Long > Short': 'Long', 'Buy': 'Long',
    'Open Short': 'Short', 'Close Short': 'Short', 'Short > Long': 'Short', 'Sell': 'Short',
    'Liquidated Isolated Short': 'Short'}
df_merged['Trade_Direction'] = df_merged['Direction'].map(direction_map)


df_merged['is_win'] = df_merged['Closed PnL'] > 0
win_rate = df_merged.groupby('Account')['is_win'].mean() * 100


df_merged['Leverage'] = df_merged['Size USD'] / df_merged['Start Position']


df_merged['Leverage'] = df_merged['Leverage'].replace([float('inf'), -float('inf')], 0).fillna(0)

leverage_stats = df_merged.groupby('Account')['Leverage'].agg(['mean', 'median', 'std']).rename(
    columns={'mean': 'Avg_Leverage', 'median': 'Median_Leverage', 'std': 'Leverage_Volatility'})


ls_counts = df_merged.groupby(['Account', 'Trade_Direction']).size().unstack(fill_value=0)
ls_counts['LS_Ratio'] = ls_counts.get('Long', 0) / (ls_counts.get('Short', 0) + 1e-6)

trader_analysis = pd.DataFrame({
    'Win_Rate_Pct': win_rate
}).join([leverage_stats, ls_counts['LS_Ratio']]).reset_index()

print("Trader Performance Analysis:")
display(trader_analysis)

Trader Performance Analysis:


Account  Win_Rate_Pct  Avg_Leverage  \
0   0x083384f897ee0f19899168e3b1bec365f52a9012     35.961236 -5.636518e+01   
1   0x23e7a7f8d14b550961925fbfdaa92f5d195ba5bd     44.271978  2.322684e+05   
2   0x271b280974205ca63b716753467d5a371de622ab     30.191651 -5.446235e+03   
3   0x28736f43f1e871e6aa8b1148d38d4994275d72c4     43.858463  8.728470e+02   
4   0x2c229d22b100a7beb69122eed721cee9b24011dd     51.991355  2.111620e+01   
5   0x3998f134d6aaa2b6a5f723806d00fd2bbbbce891     45.521472  4.791902e+05   
6   0x39cef799f8b69da1995852eea189df24eb5cae3c     32.766787  3.407368e+04   
7   0x3f9a0aadc7f04a7c9d75dc1b5a6ddd6e36486cf6     47.590361  6.461875e+03   
8   0x420ab45e0bd8863569a5efbb9c05d91f40624641     23.498695 -5.267369e+04   
9   0x430f09841d65beb3f27765503d0f850b8bce7713     48.423605  1.558817e-01   
10  0x47add9a56df66b524d5e2c1993a43cde53b6ed85     35.344524 -4.555518e+02   
11  0x4acb90e786d897ecffb614dc822eb231b4ffb9f4     48.622590 -6.721580e+02   
12  0x4f93fead39b70a1824f981a54d4e55b278e9f760     36.036392  5.285997e+04   
13  0x513b8629fe877bb581bf244e326a047b249c4ff1     40.119320 -6.105681e+05   
14  0x6d6a4b953f202f8df5bed40692e7fd865318264a     43.179487  7.900420e+07   
15  0x72743ae2822edd658c0c50608fd7c5c501b2afbd     34.591195  5.679318e+05   
16  0x72c6a4624e1dffa724e6d00d64ceae698af892a0     30.337079 -2.945641e+01   
17  0x75f7eeb85dc639d5e99c78f95393aa9a5f1170d4     81.087638  3.291530e+01   
18  0x7f4f299f74eec87806a830e3caa9afa5f2b9db8f     45.221296  4.373149e+03   
19  0x8170715b3b381dffb7062c0298972d4727a0a63b     38.274288  1.687644e+03   
20  0x8381e6d82f1affd39a336e143e081ef7620a3b7f     33.019362  1.868870e+07   
21  0x8477e447846c758f5a675856001ea72298fd9cb5     26.196826 -6.736542e+02   
22  0x92f17e8d81a944691c10e753af1b1baae1a2cd0d     28.538663  3.568189e+06   
23  0xa0feb3725a9335f49874d7cd8eaad6be45b27416     34.578661  6.053612e+05   
24  0xa520ded057a32086c40e7dd6ed4eb8efb82c00e0     57.314149  2.359598e+02   
25  0xae5eacaf9c6b9111fd53034a602c192a04e082ed     40.852575  5.770094e+00   
26  0xaf40fdc468c30116bd3307bcbf4a451a7ebf1deb     32.209738  7.514026e+04   
27  0xb1231a4a2dd02f2276fa3c5e2a2f3436e6bfed23     33.713432 -1.918961e+04   
28  0xb899e522b5715391ae1d4f137653e7906c5e2115     43.861100 -1.004753e+04   
29  0xbaaaf6571ab7d571043ff1e313a9609a10637864     46.758211  1.449112e+03   
30  0xbd5fead7180a9c139fa51a103cb6a2ce86ddb5c3     32.752745  2.991401e+04   
31  0xbee1707d6b44d4d52bfe19e41f8a828645437aab     42.823014  2.025179e+03   

    Median_Leverage  Leverage_Volatility      LS_Ratio  
0         -1.295995         1.597854e+03      0.175855  
1          0.163091         1.918856e+07      0.871076  
2         -1.677854         2.555565e+05      0.049338  
3          0.009663         7.252771e+04      1.310231  
4          0.015225         4.643968e+02   1618.499191  
5          0.005476         1.363102e+07      0.207430  
6         -0.002673         1.096175e+06      0.242729  
7          1.489461         2.454590e+04      3.151899  
8         -3.984569         1.039849e+06      0.100575  
9          0.006244         1.601456e+00      2.605341  
10        -0.019230         1.841209e+04      0.425247  
11         0.097425         4.532767e+04      9.127907  
12       330.818510         2.485201e+06      2.798091  
13         3.643252         9.847457e+07      3.163321  
14         0.014183         2.464691e+09      0.390558  
15         0.002980         2.265188e+07      0.936663  
16        -0.023726         3.677795e+02      0.163399  
17         0.000000         2.367022e+03      0.944379  
18        -6.798547         9.305850e+04      0.379646  
19        -0.001187         9.981281e+04      0.194134  
20         0.767588         8.165848e+08      1.745690  
21        -0.021780         9.547092e+04      0.114713  
22         0.006644         1.971172e+08      2.158385  
23         1.237623         2.724925e+07     12.280000  
24         0.010170         3.375294e+0

# Leverage Distribution Summary

In [27]:
df_merged['Leverage'] = (df_merged['Size USD'] / df_merged['Start Position']).abs()
df_merged['Leverage'] = df_merged['Leverage'].replace([np.inf, -np.inf], 1.0).fillna(1.0)
df_merged['Leverage_Clipped'] = df_merged['Leverage'].clip(upper=125)
print("Leverage Distribution Summary:")
print(df_merged['Leverage_Clipped'].describe())

Leverage Distribution Summary:
count    211218.000000
mean         18.597576
std          41.611862
min           0.000000
25%           0.017527
50%           0.196881
75%           2.846369
max         125.000000
Name: Leverage_Clipped, dtype: float64


# Number Of Trades Per Day

In [28]:
df_merged['date'] = pd.to_datetime(df_merged['date'])
trades_per_day = df_merged.groupby('date').size().reset_index(name='Trade_Count')
avg_daily_trades = trades_per_day['Trade_Count'].mean()
print(f"Average Trades per Day: {avg_daily_trades:.2f}")
display(trades_per_day.head())

Average Trades per Day: 440.96


date  Trade_Count
0 2023-05-01            3
1 2023-12-05            9
2 2023-12-14           11
3 2023-12-15            2
4 2023-12-16            3

# Long Short Ratio

In [29]:
ls_daily = df_merged.groupby(['date', 'Trade_Direction']).size().unstack(fill_value=0)
ls_daily['LS_Ratio'] = ls_daily.get('Long', 0) / (ls_daily.get('Short', 0) + 1e-6)
sentiment_daily = df_merged.groupby('date')['classification'].first()
ls_analysis = ls_daily.join(sentiment_daily)
display(ls_analysis[['Long', 'Short', 'LS_Ratio', 'classification']].head())

Long  Short      LS_Ratio classification
date                                                
2023-05-01     3      0  3.000000e+06          Greed
2023-12-05     7      2  3.499998e+00  Extreme Greed
2023-12-14     7      4  1.750000e+00          Greed
2023-12-15     0      2  0.000000e+00          Greed
2023-12-16     3      0  3.000000e+06          Greed

In [30]:
performance_table = df_merged.groupby('classification').agg({
    'Closed PnL': 'sum',
    'Order ID': 'count',
    'Size USD': 'mean'
}).rename(columns={'Order ID': 'Trade_Count', 'Size USD': 'Avg_Size'})
performance_table['PnL_Per_Trade'] = performance_table['Closed PnL'] / performance_table['Trade_Count']
leverage_win_table = df_merged.groupby('classification').agg({
    'Leverage_Clipped': 'median',
    'is_win': 'mean'
})
print("Insight Tables for Report")
display(performance_table)
display(leverage_win_table)

Insight Tables for Report


Closed PnL  Trade_Count     Avg_Size  PnL_Per_Trade
classification                                                       
Extreme Fear    7.391102e+05        21400  5349.731843      34.537862
Extreme Greed   2.715171e+06        39992  3112.251565      67.892861
Fear            3.357155e+06        61837  7816.109931      54.290400
Greed           2.150129e+06        50303  5736.884375      42.743559
Neutral         1.292921e+06        37686  4782.732661      34.307718

Leverage_Clipped    is_win
classification                            
Extreme Fear            0.157754  0.370607
Extreme Greed           0.156849  0.464943
Fear                    0.188998  0.420768
Greed                   0.252500  0.384828
Neutral                 0.224304  0.396991

# Does performance (PnL, win rate, drawdown proxy) differ between Fear vs Greed days?

Traders perfrom well in fear region, where they captured high Pnl and trade with more conviction, despite of low activity. On the other hand, during Greed days there is high frequency in FOMO trading that actually dilutes win rates which eventually increases the risk of massive drawdowns.

---



# Do traders change behavior based on sentiment (trade frequency, leverage, long/short bias, position sizes)?

Yes, the data confirms that there was a significant shift in behaviour as sentiment moves from Fear to Greed. During Greed and Extreme Greed, the trade frquency(peaks at 15708 orders in Extreme Greed, nearly 505x higher than in Extrem Fear) and Long/Short ratio goes up as FOMO kicks in, where as Fear region sees fewer but much bigger position sizes.Leverage also becomes aggresive during Greed.

---



# 2–3 segments :

1.   Retail FOMO traders : traders whose trade frequency goes up by 300% during Greed as compared to netural days.
2.   High Leverage Traders : Traders whose average leverage is more than 50x and have trading frequency.

---





# 3 insights backed by charts/tables.

*   Capital Efficiency vs Noise : PnL per trade is 373.88 in Fear compared to 172.85 in Extreme Greed, which means trading during fear is 2.1x more efficient.
*   Win Rate Erosion in Extreem Greed :Win rates during Extreme Greed often decline as compared to neutral.
*   Directional Bais : The Long/Short Ratio shifts from a balanced position to Neutral/Fear to a heavy Long Bias during Greed.







# 2 strategy ideas or “rules of thumb” based on your findings



*   During Fear and Extreme Fear times, traders must try to shift from high leverage speculation to high conviction trading style.
*   During Extreme Greed, implement a trade - frequency cap and exit(only for Retail FOMO traders).



# BONUS

In [34]:
pip install scikit-learn

In [59]:
from IPython.utils import process
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, classification_report
from sklearn.ensemble import RandomForestClassifier

class Trade_predictor_pipeline:
  def __init__(self):
    self.model =  RandomForestClassifier(n_estimators=100,max_depth=7, class_weight='balanced', random_state=42)
    self.scaler = StandardScaler()
    self.is_trained = False


  def preprocess_data(self,df):
    df_merged['is_win'] = (df_merged['Closed PnL'] > 0).astype(int)
    df_merged['Leverage'] = (df_merged['Size USD']/(df_merged['Start Position']+1e-6)).abs().clip(upper=125)

    daily = df_merged.groupby(['Account','date']).agg({
        'Size USD':'mean',
        'Leverage':'mean',
        'Order ID':'count',
        'Closed PnL':'sum'
    }).reset_index()

    sentiment_dummies = pd.get_dummies(df.set_index(['Account', 'date'])['classification']).groupby(level=[0,1]).first()
    daily = daily.merge(sentiment_dummies, left_on=['Account', 'date'], right_index=True)

    daily = daily.sort_values(['Account', 'date'])
    daily['target'] = (daily.groupby('Account')['Closed PnL'].shift(-1) > 0).astype(int)
    return daily.dropna()

  def train(self,df_merged):
    processed_df = self.preprocess_data(df_merged)
    features = [col for col in processed_df.columns if col not in ['Account','date','Closed PnL','target']]
    x = processed_df[features]
    y = processed_df['target']
    x_scaled = self.scaler.fit_transform(x)
    self.model.fit(x_scaled,y)
    self.is_trained = True
    return features

  def get_trader_ranking(self,df_merged):
    processed_df = self.preprocess_data(df_merged)
    features = [col for col in processed_df.columns if col not in ['Account','date','Closed PnL','target']]
    x_scaled = self.scaler.transform(processed_df[features])
    processed_df['Success_Probability'] = self.model.predict_proba(x_scaled)[:, 1]
    rankings = processed_df.groupby('Account')['Success_Probability'].mean().sort_values(ascending=False)
    return rankings

pipeline = Trade_predictor_pipeline()
feature_names = pipeline.train(df_merged)
top_traders = pipeline.get_trader_ranking(df_merged)

print("\n Top 5 Ranked Traders (Predicted Success)")
print(top_traders.head(5))



 Top 5 Ranked Traders (Predicted Success)
Account
0x28736f43f1e871e6aa8b1148d38d4994275d72c4    0.719324
0x47add9a56df66b524d5e2c1993a43cde53b6ed85    0.663766
0xbee1707d6b44d4d52bfe19e41f8a828645437aab    0.638906
0x8477e447846c758f5a675856001ea72298fd9cb5    0.633039
0x75f7eeb85dc639d5e99c78f95393aa9a5f1170d4    0.604470
Name: Success_Probability, dtype: float64


In [60]:
from sklearn.model_selection import TimeSeriesSplit
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
import seaborn as sns
import matplotlib.pyplot as plt

# 1. Use TimeSeriesSplit instead of random shuffle
# This respects the chronological order of trades
tscv = TimeSeriesSplit(n_splits=5)

processed_df = pipeline.preprocess_data(df_merged)
features = [col for col in processed_df.columns if col not in ['Account', 'date', 'Closed PnL', 'target']]
X = processed_df[features].values
y = processed_df['target'].values

# 2. Run the Walk-Forward Validation
for train_index, test_index in tscv.split(X):
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y[train_index], y[test_index]

    # Re-scale based ONLY on training data to prevent leakage
    X_train_scaled = pipeline.scaler.fit_transform(X_train)
    X_test_scaled = pipeline.scaler.transform(X_test)

    pipeline.model.fit(X_train_scaled, y_train)

# 3. Final Prediction & Metrics
y_pred = pipeline.model.predict(X_test_scaled)

print(f"--- Realistic Model Accuracy: {accuracy_score(y_test, y_pred):.2%}")
print("\n--- Final Classification Report ---")
print(classification_report(y_test, y_pred))

--- Realistic Model Accuracy: 60.26%

--- Final Classification Report ---
              precision    recall  f1-score   support

           0       0.50      0.52      0.51       156
           1       0.67      0.66      0.67       234

    accuracy                           0.60       390
   macro avg       0.59      0.59      0.59       390
weighted avg       0.60      0.60      0.60       390

